File: Leetcode/tree/684_RedundantConnection/Solution.py
class Solution(object):
    def findRedundantConnection(self, edges):
        """
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        pred = [0]*2001
        for i in xrange(0, 2001):
            pred[i] = i
        res = [-1, -1]
        for edge in edges:
            src, dest = edge[0], edge[1]
            while pred[src]!=src:
                src = pred[src]
            while pred[dest]!=dest:
                dest = pred[dest]
            if src==dest:
                res = edge
            else:
                pred[dest] = src
        return res

File: Leetcode/tree/684_RedundantConnection/Solution.java
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        int[] pred = new int[2001];
        for(int i = 0; i<2001; ++i) pred[i] = i;
        int[] res = {-1, -1};
        for(int i = 0; i<edges.length; ++i){
            int src = edges[i][0], dest = edges[i][1];
            while(src!=pred[src]) src = pred[src];
            while(dest!=pred[dest]) dest = pred[dest];
            if(src==dest){
                res = edges[i];
            }else{
                pred[dest] = src;
            }
        }
        return res;
    }
}
